### PR TITLE
Made a += b for lists do an in place add

### PR DESCRIPTION
**Pull request by Chillee**
In talks with @smessmer, we decided that it'd be better to put the logic in `list`, as optimal behavior requires knowing `.capacity()`

Results on my cpu (for the benchmark here: https://twitter.com/VahidK/status/1138674536679821312) now look like this:
```
Pytorch batch_gather took 0.018311 seconds.
Pytorch batch_gather jit took 0.013921 seconds.
Pytorch vectorized batch_gather took 0.001384 seconds.
```
Previously, `batch_gather jit` took 3x as long as `batch_gather`.

Some logic taken from https://github.com/pytorch/pytorch/pull/21690. Note that these two PR's are somewhat orthogonal. That PR handles this benchmark by looking at the alias analysis, while this PR specializes for `+=`. 

Note that we can't jit the vectorized version as we think `torch.arange` returns a float tensor.

### CODE REVIEW

File: aten/src/ATen/core/List_inl.h [hunk 1]
```diff
@@ -178,6 +178,21 @@ void ListPtr<T>::push_back(T&& value) const {
   impl_->list.push_back(detail::list_element_from<T, StorageT>(std::move(value)));
 }
 
+template<class T>
+void ListPtr<T>::append(ListPtr<T> b) const {
+  size_type neededSize = this->size() + b.size();
+  if (impl_->list.capacity() < neededSize) {
+    this->reserve(std::max(impl_->list.capacity() * 2, neededSize));
```
Comment (smessmer) [inline aten/src/ATen/core/List_inl.h]: Don't do this manually. Just ignore `capacity`/`reserve` here and use `std::vector` bulk insertion, which will do the right thing:

    impl_->list.insert(impl_->list.end(), b.impl_->list.begin(), b.impl_->list.end())

or for moving

    impl_->list.insert(v1.end(), make_move_iterator(b.impl_->list.begin()), make_move_iterator(b.impl_->list.end()));

File: torch/csrc/jit/register_prim_ops.cpp [hunk 2]
```diff
@@ -1554,6 +1554,15 @@ int listAdd(Stack& stack) {
   push(stack, std::move(ret));
   return 0;
 }
+template <class T>
+int listInplaceAdd(Stack& stack) {
+  c10::ListPtr<T> a = c10::make_list<T>();
+  c10::ListPtr<T> b = c10::make_list<T>();
+  pop(stack, a, b);
+  a.append(b);
```
Comment (smessmer) [inline torch/csrc/jit/register_prim_ops.cpp]: a.append(std::move(b))

is probably what you want instead, otherwise this shouldn't catch the optimization because the local `b` variable will keep a refcount instance.

File: torch/csrc/jit/script/compiler.cpp [hunk 3]
```diff
@@ -1632,7 +1632,15 @@ struct to_ir {
   // Get the appropriate builtin op for this augmented assignment
   // If the RHS is a tensor, return the corresponding ATen in-place op
   // If it's a list of scalars, then return the corresponding list augment op
-  Symbol getAugOp(const AugAssign& stmt, bool isTensor) {
```
Comment (smessmer) [inline torch/csrc/jit/script/compiler.cpp]: not sure what you're doing here, this seems unrelated. Can we have this in a separate PR?

**Review by smessmer — CHANGES_REQUESTED** [review]

File: torch/csrc/jit/script/compiler.cpp (hunk 3 above)
Comment (Chillee) [reply torch/csrc/jit/script/compiler.cpp]: This is what's needed to make `a += b` do the inplace add instead of desugaring it into `a = a + b`.

File: torch/csrc/jit/register_prim_ops.cpp [hunk 4]
```diff
@@ -1928,6 +1937,10 @@ RegisterOperators reg2({
           "aten::add(" decl_type "[] a, " decl_type "[] b) -> " decl_type           \
           "[]",                                                                     \
           listAdd<c_type::value_type>),                                             \
+      Operator(                                                                     \
+          "aten::add_(" decl_type "[](a!) self, " decl_type "[] b) -> " decl_type           \
```
Comment (smessmer) [inline torch/csrc/jit/register_prim_ops.cpp]: nit: line end is not aligned

File: torch/csrc/jit/register_prim_ops.cpp [hunk 5]
```diff
@@ -1554,6 +1554,15 @@ int listAdd(Stack& stack) {
   push(stack, std::move(ret));
   return 0;
 }
+template <class T>
+int listInplaceAdd(Stack& stack) {
+  c10::ListPtr<T> a = c10::make_list<T>();
```
Comment (smessmer) [inline torch/csrc/jit/register_prim_ops.cpp]: `ListPtr` got renamed to `List` recently. You'll probably have to change that.

Also, this creates an empty list just to replace it with the list from the stack two lines below. You can avoid the extra allocation by doing:

    List<T> a = pop(stack).to<List<T>>();
    List<T> b = pop(stack).to<List<T>>();

**Review by smessmer — APPROVED** [review]
looks good now. Thanks.
